[Hub Generated] Review request for Microsoft.DBforPostgreSQL to add version preview/2020-11-05-preview

tezhangmsvan opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/tezhangmsvan/azure-rest-api-specs/dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang...master/).


### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [x] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [x] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

tezhangmsvan committed:
Add files via upload

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
2 files

tezhangmsvan committed:
to json

tezhangmsvan committed:
Json resolved. Fixed most issues aware to be ARM conformant

tezhangmsvan committed:
Delete Migrations_openapi.json

tezhangmsvan committed:
Delete Migrations.json

tezhangmsvan committed:
Add files via upload

tezhangmsvan committed:
Update readme.md

tezhangmsvan committed:
Update readme.md

tezhangmsvan committed:
Update readme.md

tezhangmsvan committed:
Create custom-words.txt

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Add DMSDB in custom-words.txt

tezhangmsvan committed:
Delete custom-words.txt

tezhangmsvan committed:
Update custom-words.txt

tezhangmsvan committed:
remove sku

tezhangmsvan committed:
Update custom-words.txt

tezhangtiger committed:
add DMSDB

tezhangtiger committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
Merge branch 'dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang' of https://github.com/tezhangmsvan/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Add clouderror

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Merge branch 'master' of https://github.com/tezhangmsvan/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
remove sku

tezhangmsvan committed:
use same sku from postgresql.json

tezhangmsvan committed:
Put target state -> Post

tezhangmsvan committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
revert to put

chiragg4u commented:
Please go through the contribution checklist for ARM API Review added as part of the review and mark them completed. It will help you fix most of the issues and improve the quality of swagger and help close the PR faster. This is a requirement before review can be continued.

tezhangmsvan committed:
refactor

tezhangmsvan committed:
Dummy example and typo

tezhangmsvan committed:
remame typo

tezhangmsvan committed:
examples and summarylistresult

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
x-ms-client-flatten

tezhangmsvan committed:
x-ms-enum modelasstring

tezhangmsvan committed:
properties

tezhangmsvan committed:
array

tezhangmsvan committed:
array prettier

tezhangmsvan committed:
password null

tezhangmsvan committed:
array in create

tezhangmsvan committed:
Add description to all

tezhangmsvan committed:
MigrationCreateFullRequest allof MigrationCreateRequest

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
allof

tezhangmsvan committed:
flatten

tezhangmsvan committed:
validate examples

tezhangmsvan committed:
misc

tezhangmsvan committed:
gen examples

tezhangmsvan committed:
https://github.com/Azure/azure-rest-api-specs/pull/13555/checks?check_run_id=2506679669

tezhangmsvan committed:
update read -> get

tezhangmsvan committed:
201 created

tezhangmsvan committed:
Unify three resources

tezhangmsvan committed:
for credscan

tezhangmsvan committed:
newline at end of file

tezhangmsvan committed:
prettier playground

tezhangmsvan committed:
npx prettier --write "specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json"

tezhangmsvan committed:
array format

tezhangmsvan committed:
update examples

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+        "secretParameters": {
+          "$ref": "#/definitions/MigrationSecretParameters"
+        },
+        "targetDBServerResourceId": {
```
why are source and target necessary?  Is one not implied by the parent resource this migration is created under?

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+    "MigrationResourceProperties": {
+      "type": "object",
+      "properties": {
+        "migrationId": {
```
please indicate in the schema what fields are readonly and what is required as input

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+  },
+  "paths": {
+    "/subscriptions/{targetDBServerSubscriptionId}/resourceGroups/{targetDBServerResourceGroupName}/providers/Microsoft.DBforPostgreSQL/flexibleServers/{targetDBServerName}/migrations/{migrationId}": {
+      "put": {
```
how long does the migration take?  Should this not be an async operation?

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+      },
+      "description": "Migration resource group."
+    },
+    "MigrationSecretParameters": {
```
what are these used for and why can't Managed Resource Identity be used instead?

tezhangmsvan replied:
This is a private preview version. Will move to Managed identity in public preview.

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+      }
+    },
+    "/subscriptions/{targetDBServerSubscriptionId}/resourceGroups/{targetDBServerResourceGroupName}/providers/Microsoft.DBforPostgreSQL/flexibleServers/{targetDBServerName}/migrations": {
+      "get": {
```
this should be marked x-ms-pageable

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+            "$ref": "#/parameters/TargetDBServerNameParameter"
+          },
+          {
+            "name": "migrationListFilter",
```
any sort of query filtering should supply a rich description detailing what is supported

tezhangmsvan replied:
Changed to reference a parameter

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+            "type": "string"
+          }
+        },
+        "migrationResourceGroup": {
```
what is this resource group used for?

tezhangmsvan replied:
When the users sets startDataMigration flag to true, we will provision a Database Migration Service(DMS) under the hood an d DMS will do the heavy lifting of the migration. This resource group is used to for provisioning DMS.

tezhangmsvan commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+        "currentStatus": {
+          "$ref": "#/definitions/MigrationStatus"
+        },
+        "sourceDBServerMetadata": {
```
This is generated on the  server side. Will mark as readonly

tezhangmsvan commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:
```diff
@@ -0,0 +1,655 @@
+          "format": "date-time",
+          "type": "string"
+        },
+        "startDataMigration": {
```
Good point. Only startDataMigration and Cancel can interfere here, removing cancel. Cancel will be handled in delete API. Other flags can be set independently.

tezhangmsvan committed:
partial resolve

tezhangmsvan committed:
partial resolve

tezhangmsvan committed:
readonly and x-ms-mutability

tezhangmsvan committed:
add new version json files

tezhangmsvan committed:
Update readme.md

tezhangmsvan committed:
Merge branch 'master' into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
update

tezhangmsvan committed:
Merge branch 'dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang' of https://github.com/tezhangmsvan/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
v2

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
api version and operations

tezhangmsvan committed:
Update Migrations.json

tezhangmsvan committed:
pretty

tezhangmsvan commented:
@pilor resolved all comments and ready for another iteration of review. Thanks!

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:
```diff
@@ -0,0 +1,707 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/MigrationResource"
```
missed this on the first iteration.  Deletes should not return a body.

pilor commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:
```diff
@@ -0,0 +1,707 @@
+          },
+          {
+            "name": "parameters",
+            "in": "body",
```
GETs cannot take a body.  Any filtering should be done in the URI.  The initial iteration had it that way right?  All that was needed was a better description of the filtering that is supported

tezhangmsvan committed:
properties for patch

tezhangmsvan committed:
nameavialability

tezhangmsvan commented:
@pilor resolved comments. Thanks

tezhangmsvan committed:
migration name availability

tezhangmsvan committed:
reason

jianyexi commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:
```diff
@@ -0,0 +1,818 @@
+      },
+      "description": "Represents a migration name's availability."
+    },
+    "CloudError": {
```
Could you use  the ErrorResponse in v2/types.json instead ?

tezhangmsvan replied:
We use this object across our api's see Microsoft.DBforPostgreSQL/stable/2021-06-01/postgresql.json

tezhangmsvan committed:
Merge branch 'master' into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
add migration name

tezhangmsvan committed:
,

tezhangmsvan committed:
merge

tezhangmsvan committed:
update

tezhangmsvan committed:
merge

tezhangmsvan committed:
change version

tezhangmsvan committed:
autorest --validation --azure-validator --use=@microsoft.azure/classic-openapi-validator@latest --use=@microsoft.azure/openapi-validator@latest --tag=package-2021-06-15-privatepreview

tezhangmsvan committed:
merge

tezhangmsvan committed:
update

tezhangmsvan commented:
@jianyexi the check https://github.com/Azure/azure-rest-api-specs/pull/13555/checks?check_run_id=2835376627 
 is giving me:
<!--StartFragment-->
The schema 'StorageEditionCapability' is considered an object but without a 'type:object', please add the missing 'type:object'.Location:&nbsp;Microsoft.DBforPostgreSQL/preview/2021-06-01-preview/postgresql.json#L1161
--
<!--EndFragment-->

But if I add the type:object, it will complain that is a breaking change because the type changed. How do I resolve this?

tezhangmsvan committed:
update

tezhangmsvan committed:
update

tezhangmsvan committed:
new version postgresql.json

tezhangmsvan committed:
update

tezhangmsvan committed:
prettier

tezhangmsvan committed:
systemdata

tezhangmsvan committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
update

tezhangmsvan committed:
type object

tezhangmsvan committed:
Merge branch 'master' into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

jianyexi commented:
@tezhangmsvan  
There is only one breaking change you needs to get approval:
The new version is missing a path that was found in the old version. Was path '/providers/Microsoft.DBForPostgreSql/getPrivateDnsZoneSuffix' removed or restructured?
Old: Microsoft.DBforPostgreSQL/stable/2021-06-01/PrivateDnsZone.json#L37:5

jianyexi commented:
/azp run

jianyexi commented:
@tezhangmsvan  why you add a private preview version into the public spec repo ?

jianyexi commented:
/azp run

tezhangmsvan committed:
update

remove locatiion from rg
update error response

tezhangmsvan commented:
> @tezhangmsvan why you add a private preview version into the public spec repo ?

![image](https://user-images.githubusercontent.com/51929944/125525611-987c309f-f03b-4a30-bbe7-27e7107fe9b4.png)

tezhangmsvan commented:
> > @tezhangmsvan why you add a private preview version into the public spec repo ?
> 
> ![image](https://user-images.githubusercontent.com/51929944/125525611-987c309f-f03b-4a30-bbe7-27e7107fe9b4.png)

I think private means different things in the private repo and a private preview version. A private repo means internal to us and private preview means private preview for a customer if I understand correctly,

tezhangmsvan committed:
unify error response version

tezhangmsvan committed:
Cloud error already confroms. API update

remove location
move targetDBServerSubnetResourceId
add userAssignedIdentityResourceId

tezhangmsvan committed:
update properties for patch and fix prettier

tezhangmsvan commented:
/azp run

tezhangmsvan commented:
@jianyexi Validations passed. Can you approve? Thanks.

jianyexi commented:
> @jianyexi Jianye Xi FTE Validations passed. Can you approve? Thanks.
Since there are cross version breaking changes, could you follow the breaking change review process to get approval
https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913

tezhangmsvan commented:
> > @jianyexi Jianye Xi FTE Validations passed. Can you approve? Thanks.
> > Since there are cross version breaking changes, could you follow the breaking change review process to get approval
> > [#13555 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913)

So when I reference old API files, it tells that "XX  is considered an object but without a 'type:object', please add the missing 'type:object'". Then if I add them, it's considered a type chang and thus a breaking change. How do I resolve this?

jianyexi commented:
> > > @jianyexi
> > >             
> > >             Jianye Xi
> > >             FTE Jianye Xi FTE Validations passed. Can you approve? Thanks.
> > > Since there are cross version breaking changes, could you follow the breaking change review process to get approval
> > > [#13555 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913)
> 
> So when I reference old API files, it tells that "XX is considered an object but without a 'type:object', please add the missing 'type:object'". Then if I add them, it's considered a type chang and thus a breaking change. How do I resolve this?
Just add the missing 'type:object' , it's not breaking change now

jianyexi commented:
I mean the below breaking change need get approval, not the breaking change caused by adding missing type object:
![image](https://user-images.githubusercontent.com/59603451/126055993-8268b72f-1955-419d-b994-cacd6dc33379.png)

tezhangmsvan commented:
> I mean the below breaking change need get approval, not the breaking change caused by adding missing type object:
> ![image](https://user-images.githubusercontent.com/59603451/126055993-8268b72f-1955-419d-b994-cacd6dc33379.png)

That was outdated with my latest changes. Please see my latest changes and checks.

tezhangmsvan committed:
update

tezhangmsvan committed:
Service error

tezhangmsvan commented:
@jianyexi Swagger LintDiff  complains about errors in our exsiting API. How do I resolve this? If I edit the exist APIs by adding the object type, it'll tell me it's a breaking change.

tezhangmsvan committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang

tezhangmsvan committed:
type; object, duplicate version

tezhangmsvan committed:
pretty and unused example

tezhangmsvan committed:
remove privatezone

tezhangmsvan committed:
remvoe db

tezhangmsvan committed:
Sql -> SQL

tezhangmsvan commented:
@jianye operations get was clearly implemented, why is it complaining?
Operations API must be implemented for '/providers/Microsoft.DBForPostgreSQL/operations'.
Location: Microsoft.DBforPostgreSQL/preview/2021-06-15-privatepreview/postgresql.json#L36

tezhangmsvan committed:
casing

tezhangmsvan commented:
@jianyexi looks all are passing. Can you approve?

jianyexi commented:
> @jianyexi Jianye Xi FTE looks all are passing. Can you approve?
> 
> So I added the "type : obejct" to conform, but then the cross version breaking changes consider this a type change and breaking? How do I resolve this?

It's not breaking change now, are you ready for merging ?

tezhangmsvan commented:
> > @jianyexi Jianye Xi FTE looks all are passing. Can you approve?
> > So I added the "type : obejct" to conform, but then the cross version breaking changes consider this a type change and breaking? How do I resolve this?
> 
> It's not breaking change now, are you ready for merging ?

Yes, we are ready to merge. Thanks.

tezhangmsvan commented:
@jianyexi ready to merge. Thanks
